diff --git a/src/protocols/core/Compositor.cpp b/src/protocols/core/Compositor.cpp
index 2d417236725..516b4a4990e 100644
--- a/src/protocols/core/Compositor.cpp
+++ b/src/protocols/core/Compositor.cpp
@@ -675,11 +675,11 @@ PImageDescription CWLSurfaceResource::getPreferredImageDescription() {
 
     auto parent = m_self;
     if (parent->m_role->role() == SURFACE_ROLE_SUBSURFACE) {
-        auto subsurface = sc<CSubsurfaceRole*>(parent->m_role.get())->m_subsurface.lock();
-        parent          = subsurface->t1Parent();
+        const auto subsurface = sc<CSubsurfaceRole*>(parent->m_role.get())->m_subsurface.lock();
+        parent                = subsurface ? subsurface->t1Parent() : nullptr;
     }
     PHLMONITORREF monitor;
-    if (parent->m_enteredOutputs.size() == 1)
+    if (parent && parent->m_enteredOutputs.size() == 1)
         monitor = parent->m_enteredOutputs[0];
     else if (m_hlSurface.valid() && WINDOW)
         monitor = WINDOW->m_monitor;
diff --git a/src/protocols/core/Subcompositor.cpp b/src/protocols/core/Subcompositor.cpp
index d8a3e8608de..8573d30e598 100644
--- a/src/protocols/core/Subcompositor.cpp
+++ b/src/protocols/core/Subcompositor.cpp
@@ -125,9 +125,13 @@ Vector2D CWLSubsurfaceResource::posRelativeToParent() {
     // surfaces we've visited and if we hit a surface we've visited we bail out.
     std::vector<SP<CWLSurfaceResource>> surfacesVisited;
 
-    while (surf->m_role->role() == SURFACE_ROLE_SUBSURFACE && std::ranges::find_if(surfacesVisited, [surf](const auto& other) { return surf == other; }) == surfacesVisited.end()) {
+    while (surf && surf->m_role->role() == SURFACE_ROLE_SUBSURFACE &&
+           std::ranges::find_if(surfacesVisited, [surf](const auto& other) { return surf == other; }) == surfacesVisited.end()) {
         surfacesVisited.emplace_back(surf);
-        auto subsurface = sc<CSubsurfaceRole*>(m_parent->m_role.get())->m_subsurface.lock();
+        const auto subsurface = sc<CSubsurfaceRole*>(surf->m_role.get())->m_subsurface.lock();
+        if (!subsurface)
+            break;
+
         pos += subsurface->m_position;
         surf = subsurface->m_parent.lock();
     }
@@ -142,10 +146,14 @@ SP<CWLSurfaceResource> CWLSubsurfaceResource::t1Parent() {
     SP<CWLSurfaceResource>              surf = m_parent.lock();
     std::vector<SP<CWLSurfaceResource>> surfacesVisited;
 
-    while (surf->m_role->role() == SURFACE_ROLE_SUBSURFACE && std::ranges::find_if(surfacesVisited, [surf](const auto& other) { return surf == other; }) == surfacesVisited.end()) {
+    while (surf && surf->m_role->role() == SURFACE_ROLE_SUBSURFACE &&
+           std::ranges::find_if(surfacesVisited, [surf](const auto& other) { return surf == other; }) == surfacesVisited.end()) {
         surfacesVisited.emplace_back(surf);
-        auto subsurface = sc<CSubsurfaceRole*>(m_parent->m_role.get())->m_subsurface.lock();
-        surf            = subsurface->m_parent.lock();
+        const auto subsurface = sc<CSubsurfaceRole*>(surf->m_role.get())->m_subsurface.lock();
+        if (!subsurface)
+            break;
+
+        surf = subsurface->m_parent.lock();
     }
     return surf;
 }
@@ -174,8 +182,13 @@ CWLSubcompositorResource::CWLSubcompositorResource(SP<CWlSubcompositor> resource
         SP<CWLSurfaceResource> t1Parent = nullptr;
 
         if (PARENT->m_role->role() == SURFACE_ROLE_SUBSURFACE) {
-            auto subsurface = sc<CSubsurfaceRole*>(PARENT->m_role.get())->m_subsurface.lock();
-            t1Parent        = subsurface->t1Parent();
+            const auto subsurface = sc<CSubsurfaceRole*>(PARENT->m_role.get())->m_subsurface.lock();
+            if UNLIKELY (!subsurface) {
+                r->error(WL_SUBCOMPOSITOR_ERROR_BAD_PARENT, "Parent has no valid subsurface role");
+                return;
+            }
+
+            t1Parent = subsurface->t1Parent();
         } else
             t1Parent = PARENT;
 
